test/e2e/azure_logcollector.go

```diff
@@ -0,0 +1,192 @@
+// +build e2e
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"golang.org/x/crypto/ssh"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	kinderrors "sigs.k8s.io/kind/pkg/errors"
+)
+
+// AzureLogCollector collects logs from a CAPZ workload cluster.
+type AzureLogCollector struct{}
+
+// CollectMachineLog collects logs from a machine.
+func (k AzureLogCollector) CollectMachineLog(ctx context.Context,
+	managementClusterClient client.Client, m *clusterv1.Machine, outputPath string) error {
+	cluster, err := util.GetClusterFromMetadata(ctx, managementClusterClient, m.ObjectMeta)
+	if err != nil {
+		return err
+	}
+	controlPlaneEndpoint := cluster.Spec.ControlPlaneEndpoint.Host
+	hostname := m.Spec.InfrastructureRef.Name
+	execToPathFn := func(outputFileName, command string, args ...string) func() error {
+		return func() error {
+			f, err := fileOnHost(filepath.Join(outputPath, outputFileName))
+			if err != nil {
+				return err
+			}
+			defer f.Close()
+			return execOnHost(controlPlaneEndpoint, hostname, f, command, args...)
+		}
+	}
+
+	return kinderrors.AggregateConcurrent([]func() error{
+		execToPathFn(
+			"journal.log",
+			"journalctl", "--no-pager", "--output=short-precise",
+		),
+		execToPathFn(
+			"kern.log",
+			"journalctl", "--no-pager", "--output=short-precise", "-k",
+		),
+		execToPathFn(
+			"kubelet-version.txt",
+			"kubelet", "--version",
+		),
+		execToPathFn(
+			"kubelet.log",
+			"journalctl", "--no-pager", "--output=short-precise", "-u", "kubelet.service",
+		),
+		execToPathFn(
+			"containerd.log",
+			"journalctl", "--no-pager", "--output=short-precise", "-u", "containerd.service",
+		),
+		execToPathFn(
+			"cloud-init.log",
+			"cat", "/var/log/cloud-init.log",
+		),
+		execToPathFn(
+			"cloud-init-output.log",
+			"cat", "/var/log/cloud-init-output.log",
+		),
+	})
+}
+
+// execOnHost runs the specified command directly on a node's host, using an SSH connection
+// proxied through a control plane host.
+func execOnHost(controlPlaneEndpoint, hostname string, f *os.File, command string,
+	args ...string) error {
+	config, err := newSSHConfig()
+	if err != nil {
+		return err
+	}
+	port := "22"
```

Comment:
one thing I just realized, this won't work for VMSS (I think VMSS ssh port is 2201?). We don't support VMSS control planes at the moment but it's worth making this parametrizable for the future or at least leaving a note here

Reply:
it's 50001*